### ISSUE TITLE
Merging persons improvement

### ISSUE THREAD

**Issue by EDsCODE** [PostHog/plugin-server#329]
***Edit: Important comment below that might be source of duplication**

**Current situation:**

In `src/worker/ingestion/process-event.ts` we have a function that handles aliasing or, in other words, identifying the user. This logic handles how users are merged depending on the payload that comes with the identify calls. We rely on `alias` or `$anon_distinct_id` to discover previously stored versions of the person in question. If possible, this function will then merge those users together as we want to ensure that we're keeping one source of truth per user.

The issue here is that we're relying on `alias` and `$anon_distinct_id` which is locally persisted data by the client library (found [here](https://github.com/PostHog/posthog-js/blob/cbf631811a4014b8edb4b8172aa0454773118780/src/posthog-core.js#L831) and [here](https://github.com/PostHog/posthog-js/blob/cbf631811a4014b8edb4b8172aa0454773118780/src/posthog-core.js#L921)). As a result, we will have moments where the local persistence is cleared and our aliasing function will not be able to associate the current user being identified to any preexisting user. The function will then decide to insert the user as a completely new person and never automatically merge this person with previous instances. [Example](https://metabase.posthog.net/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJTRUxFQ1QgKiBmcm9tIHBvc3Rob2dfcGVyc29uIHdoZXJlIHRlYW1faWQgPSAyIGFuZCBwcm9wZXJ0aWVzIC0-ICdlbWFpbCcgPSAnXCJkYW5pZWxAbWFkYmF0dGVyLmJpelwiJyIsInRlbXBsYXRlLXRhZ3MiOnt9fSwiZGF0YWJhc2UiOjN9LCJkaXNwbGF5IjoidGFibGUiLCJ2aXN1YWxpemF0aW9uX3NldHRpbmdzIjp7fX0=)

**Consequences:** fairly straightforward but listing for completeness
- inconclusive results on analytic queries as there will be double counting of a single user
- confusing copies when browsing person data
- unnecessary storage of multiple copies of a person (inconsequential now but effects could grow as we become more heavily used)

**Goal:**
Have a solution that effectively handles merging of persons such that there is only ever a single row per person being tracked for a product/team

**Solution:**
- Maintain our generated distinct_id scheme consistently. Add a `custom_identifier` column to posthog_person table that will store the identifier that is passed in through identify if a developer chooses to do so (ex: email). This will be done instead of taking over distinct_id 
        1.  By default, our system will still use the local persistence distinct_id to try to merge users if possible/if the alias function receives a distinct_id from the client
        2. If the above fails, the alias function will check to see if there are any preexisting users that can be merged on `custom_identifier`. So if there are preexisting `eric@posthog.com`'s, it will now fetch those rows and merge it altogether so there will be no remaining copies

**I'm going to work on this and open a PR anyway so that it's clearer how I'm trying to fix this issue however also opening a ticket to start because this is a significant modification to how our ingestion pipeline handles users. Everyone should make sure this makes sense.

**Comment by kpthatsme** [comment]
Thanks for this write up @EDsCODE it sounds pretty good to me based on the convo from earlier! 

I have questions around how we're going to be determining and setting the custom identifier (i.e. is this dynamic, set by a team, etc) but I think those will be clearer from the code.

**Comment by EDsCODE** [comment]
Right now, when a team passes in their own identifier in the `.identify` function we use that instead of our generated distinct_id. So in my solution, we would maintain the distinct_id that we generate for uniformity in our system, and then save the identifier they pass in a separate column on posthog_person table. In other words, our system will attempt to merge based on distinct_ids and on the user provided identifier. 

The user provided identifier should always be accessible because this would be the email/id that a product uses to identify their users, so anytime the user is logged in, they would be passing this id into posthog. This is pretty much happening now which is why we get several rows of a user with the same email. In this solution, our system would merge on users with that email to hopefully ensure there's always a single row.

Important to note is that we'll also need to make sure we can keep this in sync on clickhouse so that the actual analytic querying is proper too which I think should be fine. I'll provide more thoughts here as I figure it out

**Comment by EDsCODE** [comment]
Didn't realize our identify call strictly requires an ID arg. The above still works but there's no need to handle generated distinct_ids and uniqueid separately. We would just make sure to store the provided unique ID in the person table.

I also discovered some further strangeness. The duplicated persons [are unique per org/project](https://metabase.posthog.net/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJTRUxFQ1QgKiBGUk9NIHBvc3Rob2dfcGVyc29uIHdoZXJlIHRlYW1faWQgPSAyIGFuZCBwcm9wZXJ0aWVzIC0-ICdlbWFpbCcgPSAnXCJkYW5pZWxAbWFkYmF0dGVyLmJpelwiJ1xuIiwidGVtcGxhdGUtdGFncyI6e319LCJkYXRhYmFzZSI6M30sImRpc3BsYXkiOiJ0YWJsZSIsInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9fQ==)

**Comment by mariusandra** [comment]
Hey, you've obviously thought about it before posting the issue, hence I guess the approach makes sense. Still I'd like to challenge a bit to ask if it's actually us who are incorrectly using posthog here? It's always easy to build new features, but perhaps just by changing the API somehow we can avoid this?

It just looks like the user was identified with a [different distinct id](https://github.com/PostHog/posthog/blob/ae6b8eef904ee6bcfe9bd024e20a9b6c9e3c1817/posthog/models/user.py#L28) in all of these cases? Maybe the solution here is that [we call](https://github.com/PostHog/posthog/blob/76f7c1fd543330976bf18f43c29b81662c598a0d/frontend/src/scenes/userLogic.tsx#L75) `posthog.identify("email-of@user.com")` instead of using that random autogenerated distinct_id that's unique per posthog installation?

If this was discussed before, then I'm sorry for missing that context :).

That said, I have nothing really against keeping distinct ids in the format of randomly generated UUIDs and having a separate field to track the user-supplied identifiers. But even then, this needs to be thought through. What happens for example if we have more than one identifier? E.g. if the user calls `alias` and adds more custom identifiers than the field in `person` can handle? Won't the increased complexity bite us at some point down the line? E.g. in analytics queries, in the postgres<->clickhouse synchronisation, etc...?

To reiterate, I'm not against this, just trying to get a better understanding.

**Comment by EDsCODE** [comment]
Absolutely agree. In this case, I'm quite against making changes to the ingestion flow if it's not needed because it will cause more pain and bloat if it's solving a problem that doesn't actually exist. I think you're right about the different distinct_ids. The last thing I noticed last night was that it seems the duplicate persons are tied to different orgs. Kunal and I spot checked a bunch of different users we noticed were duplicated and all duplicates were associated with different orgs. So my current hunch is our current ingestion might be working as expected but there's something in org creation/joining logic that will force create a new person. Currently looking through that logic as I'm unfamiliar with all the effects of creating an org
